junit-platform-engine/src/main/java/org/junit/platform/engine/support/descriptor/FilePosition.java

```diff
@@ -77,7 +80,7 @@ public static FilePosition from(int line, int column) {
 	 * @see #from(int)
 	 * @see #from(int, int)
 	 */
-	public static Optional<FilePosition> fromQuery(String query) {
+	public static Optional<org.junit.platform.engine.discovery.FilePosition> fromQuery(String query) {
```

Comment:
Returning the super class here will break existing callers.

Reply:
This is slightly embarrassing, but when I change it to `FilePosition` it refuses to compile:
```
> Task :junit-platform-engine:compileJava FAILED
/junit5/junit-platform-engine/src/main/java/org/junit/platform/engine/support/descriptor/FilePosition.java:85: error: fromQuery(String) in org.junit.platform.engine.support.descriptor.FilePosition cannot hide fromQuery(String) in org.junit.platform.engine.discovery.FilePosition
        public static Optional<FilePosition> fromQuery(String query) {
                                             ^
  return type Optional<org.junit.platform.engine.support.descriptor.FilePosition> is not compatible with Optional<org.junit.platform.engine.discovery.FilePosition>
1 error
``` 
I fail to understand why this is only an issue for this static method and not the other two static `from` methods that take one or more `int` arguments. Any suggestions on what to do here?

Reply:
I've briefly explored using `Optional<? extends FilePosition>` in the superclass `FilePosition`, but that leaks out into usage of that method regrettably. :\

Reply:
A method can hide a method in the parent class if it follows he Liskov Substitution Principle. So a subclass must return a more specific type then the parent class. For generics this inversed.

Suppose it would be possible to do extend a class and return a more specialized generic:

```java
class Parent {
    public List<Number> funX() {
        return new ArrayList<>();
    }
}

class Child extends Parent {
    public List<Double> funX() {
        return new SpecializedDoubleList<>();
    }
}
```

Then by invoking `funX` on the parent type we could get a list that is only intended to contain `Double` values and add an `Integer` to it.

```java
Parent p = new Child();
List<Number> numbers = p.funX();
numbers.add(Integer.valueOf(1));
```
So the problem is that generic type is too specific. It specifies both the elements that can be added to the list (as a consumer) as well as those that can be taken from it (as a producer). By changing the parent to `extends Number` we can specify that the list contains an unknown type of element, but if you were to taken elements from it, it would at-least extend `Number`.

```java
class Parent {
    public List<? extends Number> funX() {
        return new ArrayList<>();
    }
}
```

Reply:
> I've briefly explored using Optional<? extends FilePosition> in the superclass FilePosition, but that leaks out into usage of that method regrettably. :\

It isn't pretty but I think that is okay. 

It is possible to assign `List<Number>` to `List<? extends Number>` so once the old `FilePosition` and the `extends Number` are removed it shouldn't break anything.

Reply:
Applied in https://github.com/junit-team/junit5/pull/2253/commits/decb66931b828437d53497fcf091ba3209f36774 as mentioned above; probably best for @marcphilipp to have another close look :)